Implementation of co-process hooks

tgross opened the pull request:
For https://github.com/joyent/containerpilot/issues/175, inspired by discussions in https://github.com/joyent/containerpilot/issues/162.

This PR implements user-defined hooks for coprocesses: processes that are started but remain resident rather than exiting immediately. These can be used to run local Consul agents, daemons like confd or consul-template, log shippers in non-Docker environments, etc.

justenwalker commented on coprocesses/coprocess.go:
```diff
@@ -0,0 +1,105 @@
+func (coprocess *Coprocess) Start() {
+	log.Debugf("coprocess[%s].Start", coprocess.Name)
+	fields := log.Fields{"process": "coprocess", "coprocess": coprocess.Name}
+	stdout := utils.NewLogWriter(fields, log.InfoLevel)
```
Using LogWriter will take care of that weird interleaving problem I mentioned before.  Good catch.

justenwalker commented on coprocesses/coprocess.go:
```diff
@@ -0,0 +1,105 @@
+	defer coprocess.closeLogs()
+
+	for {
+		if coprocess.Restarts < 0 {
```
This logic might have a problem.  If the Restarts counter is not reset after a successful start, then a process that sporadically dies and is restarted successfully may eventually fail to restart in the future.  

Defining success here could be tricky though... Some options:

1. coprocesses have health checks. success means that the health check passed.
2. success means the coprocess is still running after some delay

tgross replied:
> then a process that sporadically dies and is restarted successfully may eventually fail to restart in the future.

If you've got something that flaky then I you probably want it to always restart, I guess? Maybe we should make the Restarts optional? How would you recommend we handle that from a configuration standpoint. My understanding is that we end up with a value of 0 if it's omitted, which we could special case but then we wouldn't have a way to say "never restart"

justenwalker replied:
In the first pass you mentioned that a dying coprocess kills the container - Which is probably preferable to restarting it a finite number of times. The traditional way is to treat `-1` special like "always restart" and then "0" means what it says on the tin. 

However, I hate doing that. I'd try to avoid doing the whole special values case if we can just put another config in.  I'd take a page out of the [upstart](http://upstart.ubuntu.com/cookbook/#respawn-limit) cook-book and make two configs:

- restart: true / fase
- restartLimit: 1+ | "unlimited"

Since we are doing raw configs, we can detect the type of `restartLimit` and actually get a nice self-documenting value.

justenwalker replied:
We could also make the restartLimit more human readable and just always accept a string argument

- `restartLimit: "3"`
- `restartLimit: "3 wait 5s"`
- `restartLimit: "unlimited"`

tgross replied:
If we're going to accept a multi-type value for `restartLimit`, we could just default to no restart and avoid having the extra `restart: true|false`

tgross replied:
Ok, I've taken another pass at that.

tgross commented on coprocesses/coprocess.go:
```diff
@@ -0,0 +1,148 @@
+			coprocess.restartsRemain = t
+			break
+		} else {
+			// otherwise ignore it; can't fallthrough on type switch
```
Actually these are the zero-values for these struct fields... I kinda like that it's explicit here but maybe we should omit it and then have the `default` type case raise a parse error?

justenwalker replied:
Should probably check for field omission before type asserting - and then raise an error in the default case as you said.

tgross replied:
Sounds good on all the above.

justenwalker commented on coprocesses/coprocess.go:
```diff
@@ -0,0 +1,148 @@
+			coprocess.restartLimit = unlimitedRestarts
+			coprocess.restartsRemain = unlimitedRestarts
+		} else {
+			return fmt.Errorf("`restartLimit` field only takes positive integers or `unlimited`")
```
I'd also put a case in for "never" to make it easy to generate/template the config. Omission as the **only way** to get the desired behavior is inconvenient.

justenwalker commented on coprocesses/coprocess.go:
```diff
@@ -0,0 +1,148 @@
+		} else {
+			return fmt.Errorf("`restartLimit` field only takes positive integers or `unlimited`")
+		}
+	case int:
```
Also, for consistency - we should support strings with integer contents like we do for stopTimeout, etc...

tgross replied:
Oh, right. I'd forgotten about that.

tgross commented:
I've updated the config parser to accept `"never"` as well as strings-that-parse-as-ints. And I've added a bunch of table-based tests for the restart parsing and behavior.

justenwalker commented:
Another nitpick - probably my fault too:

```
restartLimit: never
restartLimit: unlimited
restartLimit: 4
```

Doesn't read well to me.  especially never (even though we should have it). Not really a huge deal, but maybe there is a better name for `never` or for `restartLimit`

tgross commented:
Maybe calling the field `restarts` would be better? That way it's grammatical either way.

```
restarts: never
restarts: unlimited
restarts: 4
```

justenwalker commented:
Yeah that works for me

tgross commented:
Done on the field rename. That reads a lot nicer in the code too.

Next on the list are some integration tests and documentation updates.

tgross commented:
I've worked up an integration test that exercises the restart behavior as well as ensures that the restart limit is being reset after a SIGHUP to reload the configuration. I've also added some more testing of the raw config load for this package, because mapstructure had a couple little surprises for me as described in https://github.com/joyent/containerpilot/pull/176/commits/771f25c6cf2cfb2a58c000f5e0e0a1cdb850831d

Next up is to document it all.

tgross commented:
@misterbisson I've pushed up a bunch of associated documentation changes into this PR now too.

tgross commented:
Just pushed a big ol' rebase so that I've got the implementation (+ unit tests), the integration tests, and the documentation all as clean commits.

tgross commented:
@justenwalker any last comments on this one?

justenwalker commented:
@tgross Just that you've done an excellent job. LGTM.

adamlc commented:
Massive thanks for this, very useful!

For those that have apps that need to wait for coprocesses to do something before startong I found this tool called [wait-on](https://github.com/jeffbski/wait-on) that makes this really easy. Not sure if it'll affect signals, but I'm using directly in my node app anyway :)
